crates/net/src/sim2h_worker.rs

```diff
@@ -209,58 +232,65 @@ impl Sim2hWorker {
-                    )
-                });
-            let signed_wire_message = SignedWireMessage::new(
-                message.clone(),
-                Provenance::new(self.agent_id.clone(), signature.into()),
+        }
+        let message = &buffered_message.wire_message;
+        debug!("WireMessage: preparing to send {:?}", message);
+        let payload: String = message.clone().into();
+        let signature = self
+            .conductor_api
+            .execute(payload.clone(), CryptoMethod::Sign)
+            .unwrap_or_else(|e| {
+                panic!(
+                    "Couldn't sign wire message in sim2h worker: payload={}, error={:?}",
+                    payload, e
+                )
+            });
+        let payload: Opaque = payload.into();
+        let signed_wire_message = SignedWireMessage::new(
+            payload.clone(),
+            Provenance::new(self.agent_id.clone(), signature.into()),
+        );
+        let to_send: Opaque = signed_wire_message.into();
+
+        // safe to unwrap because we check connection_ready() above
+        if let Err(e) = self
+            .connection
+            .as_mut()
+            .unwrap()
+            .write(to_send.to_vec().into())
+        {
+            error!(
+                "TransportError trying to send message to sim2h server: {:?}",
+                e
             );
-            let to_send: Opaque = signed_wire_message.into();
-            // safe to unwrap because we check connection_ready() above
-            if let Err(e) = self
-                .connection
-                .as_mut()
-                .unwrap()
-                .write(to_send.to_vec().into())
-            {
-                error!(
-                    "TransportError trying to send message to sim2h server: {:?}",
-                    e
-                );
-                self.connection = None;
-                self.check_reconnect();
-                return did_something;
-            }
-            debug!("WireMessage: dequeuing sent message {:?}", message);
-            // if we made it here, we successfully sent the first message
-            // we can remove it from the outgoing buffer queue
-            self.outgoing_message_buffer.remove(0);
+            self.connection = None;
+            self.check_reconnect();
+            return true;
         }
+        #[allow(deprecated)]
+        let mut hasher = SipHasher::new();
+        payload.hash(&mut hasher);
+        buffered_message.hash = hasher.finish();
+        buffered_message.last_sent = Some(Instant::now());
+        true
     }
 
     /// if we re-connected, we may need to send a join first,
     /// before other queued messages
     fn prepend_wire_message(&mut self, message: WireMessage) -> NetResult<()> {
         debug!("WireMessage: queueing {:?}", message);
-        self.outgoing_message_buffer.insert(0, message);
+        self.outgoing_message_buffer.insert(0, message.into());
```

Comment:
Do we need to set the `last_sent` for the head of the queue to `None` here - so we don't wait on it after this prepended join is processed?

Reply:
That happens here: https://github.com/holochain/holochain-rust/pull/2120/files#diff-03e64f941c57272f7ef1b84b07ed2005R66

Reply:
```rust
impl From<WireMessage> for BufferedMessage {
    fn from(wire_message: WireMessage) -> BufferedMessage {
        BufferedMessage {
            wire_message,
            hash: 0,
            last_sent: None,
        }
    }
}
```

Reply:
no - I mean the previous head of the outgoing queue (item index 0) which may still be pending an Ack - if we prepend a JoinSpace message and that join space is successful, the next message may still have a not-None `last_sent` in it that should be cleared, right?

Reply:
Right! Good catch!